hip_data_tools/oracle/mysql.py

```diff
@@ -0,0 +1,172 @@
+"""
+Utility to read and write from and to mysql
+"""
+from builtins import str
+from contextlib import closing
+from dataclasses import dataclass
+from typing import Optional, Dict, Tuple, Any
+
+import MySQLdb
+import pandas.io.sql as sql
+from MySQLdb.cursors import BaseCursor
+from pandas import DataFrame
+
+from hip_data_tools.common import SecretsManager, KeyValueSource, ENVIRONMENT, LOG
+
+
+def prepare_upsert_query(table: str, primary_keys: Dict[str, Any],
+                         data: Dict[str, Any]) -> (str, Tuple):
+    pk_columns = primary_keys.keys()
+    data_columns = data.keys()
+    pk_col_list = ", ".join(pk_columns)
+    data_col_list = ", ".join(data_columns)
+    parameter_list = ",".join(["%s" for col in pk_columns] + ["%s" for col in data_columns])
+    update_list = ", ".join([f"{col}=%s" for col in data_columns])
+    values = tuple(list(primary_keys.values()) + list(data.values()))
```

Comment:
🤔 could it happen that we get a mix between which values relate to which keys? Does the dict structure have an internal order to it that is reliable?

Reply:
Yes. Dist sequence is reliable

Reply:
It seems to be from Python 3.7 but not necessarily prior to that?
https://stackoverflow.com/questions/5629023/the-order-of-keys-in-dictionaries

Can you provide a link to the sequence order of keys and values being reliably the same? @arihantsurana

Reply:
the arbitrary nature of sequencing in a dict is referenced in relation to the order of keys and values before and after addition to a dict. However, once inside a dict the keys() will return an arbitrary sequence, but the values(0 will follow the same sequence, even though its arbitrary.
I believe so, because keys and values are calculated from same dict view https://docs.python.org/3.8/library/stdtypes.html#dictionary-view-objects

see example on page
```
>>> # keys and values are iterated over in the same order (insertion order)
>>> list(keys)
['eggs', 'sausage', 'bacon', 'spam']
>>> list(values)
[2, 1, 1, 500]

```